### ISSUE TITLE
How to set  pyenv's  virtualenv?

### ISSUE THREAD

**Issue by nanounanue** [dzop/emacs-jupyter#93]
I am trying to run 

```org
#+BEGIN_EXAMPLE jupyter-python :session test
import pandas as pd
#+END_EXAMPLE
```

I am using `pyenv` to set a `virtualenv`, but that's is no picked by default. The file is inside a directory with a `.python-version`

The error I am getting is:

```
Debugger entered--Lisp error: (invalid-slot-name "#<jupyter-org-client jupyter-org-client-1558196582..." ioloop)
  signal(invalid-slot-name ("#<jupyter-org-client jupyter-org-client-1558196582..." ioloop))
  slot-missing(#<jupyter-org-client jupyter-org-client-155819658260> ioloop oref)
  slot-value(#<jupyter-org-client jupyter-org-client-155819658260> ioloop)
  #f(compiled-function (#<jupyter-org-client jupyter-org-client-155819658260>)
  apply(#f(compiled-function 
  #f(compiled-function (&rest args) #<bytecode 0x155815889515>)(#<jupyter-org-client jupyter-org-client-155819658260>)
  apply(#f(compiled-function (&rest args) #<bytecode 0x155815889515>) #<jupyter-org-client jupyter-org-client-155819658260> nil)
  jupyter-start-channels(#<jupyter-org-client jupyter-org-client-155819658260>)
  jupyter-start-new-kernel("python3" jupyter-org-client)
  jupyter-run-repl("python3" nil nil jupyter-org-client)
....
```

**Comment by dzop** [comment]
I can't get much from the error. Can you show the output of 
`(car (jupyter-get-kernelspec "python3"))`? The problem could be related to the kernelspec pointing to a python using an absolute path.

I'm not familiar with `pyenv` so I'm not sure what the significance of `.python-version` is, but from my understanding `pyenv` modifies the `PATH` shell variable. So you will have to ensure that

1. `(car (jupyter-get-kernelspec "python3"))` doesn't point to an absolute path
2. The `pyenv` `python` is found before any other python by ensuring the `PATH` in the 
`process-environment` variable is set to an appropriate value.

**Comment by nanounanue** [comment]
I will do as soon as I have access to my computer

**Comment by dzop** [comment]
@nanounanue Where you able to get this working?

**Comment by sjkuntz** [comment]
I am also trying to get environments to work, but I'm using `pyvenv` and `conda`. I have an environment called `euler` that has some modules I need to use in an `org` file, so I activate that environment with `pyvenv-workon`, and run the following code block

```
#+begin_src jupyter-python :session py
import euler
#+end_src
```

but I get a ``ModuleNotFoundError``. So I'm guessing ``emacs-jupyter`` isn't opening the kernel I want. I tried what you suggested above

1. ``(car (jupyter-get-kernelspec "python3"))`` gives ``/home/sk/anaconda3/share/jupyter/kernels/python3``
2. ``process-environment`` is set to ``VIRTUAL_ENV=/home/sk/anaconda3/envs/euler`` and ``PATH=/home/sk/anaconda3/envs/euler/bin:...``

If I can get the kernelspec to point to ``VIRTUAL_ENV/share/jupyter/kernels/python3`` I think that will fix my problem, but I wanted to check if there is a better solution before trying to hack it.

I had also tried passing the kernel file directly into the session variable. Is it possible to start a session with a specified kernel? This still wouldn't be an adequate solution for me - I'd prefer integration with the existing environment managers.

**Comment by dzop** [comment]
> `(car (jupyter-get-kernelspec "python3"))` gives `/home/sk/anaconda3/share/jupyter/kernels/python3`

This points to the directory of the kernelspec. We get the kernelspecs by calling the `jupyter kernelspec list` shell command and we cache the results of that call so when you switch virtualenvs the cached kernelspecs don't get updated. You can refresh the cached kernelspecs by evaluating `(jupyter-available-kernelspecs t)` after switching virtualenvs. This may solve your problems.

Another option is to create a kernelspec in one of the directories searched by Jupyter (https://jupyter.readthedocs.io/en/latest/projects/jupyter-directories.html#envvar-JUPYTER_PATH) that is outside of you virtualenv installation and that uses a relative file name for the Python program.

Your `python3` kernelspecs probably all use absolute file names as the program name. You can confirm this by checking the first element of the vector returned by 
`(plist-get (cdr (jupyter-get-kernelspec "python3")) :argv)`. If you make a new kernelspec that uses just `python3` as the program name then, when starting the kernel process in Emacs, the `exec-path` variable will be looked at to find the actual program and will end up using the python in your virtualenv.

> Is it possible to start a session with a specified kernel?

```#+begin_src jupyter-python :session py :kernel python3```

You can't specify a directory though. Whatever kernelspecs returned by the shell command 
`jupyter kernelspec list` is what you can set the `:kernel` header argument to.

**Comment by sjkuntz** [comment]
Thank you, adding ``(add-hook 'pyvenv-post-activate-hooks (lambda() (jupyter-available-kernelspecs t)))`` to my init file worked. Restarting the kernel doesn't activate the new environment, only closing the session and starting a new session does, but that's something I can live with.

**Comment by magonreal** [comment]
I have the same problem with my spacemacs and pyenv. I am using miniconda3 and my virtualenv is called `ahc`. When I try to evaluate source code in org-mode with `jupyter-python` I also get the message 
```
executing Jupyter-Python code block...
Starting python3 kernel process...done
user-error: Modules are not supported
```

If I check 
`(car (jupyter-get-kernelspec "python3"))`, I get
`"/home/user/.pyenv/versions/miniconda3-latest/envs/ahc/share/jupyter/kernels/python3"`

For `(plist-get (cdr (jupyter-get-kernelspec "python3")) :argv)`, I get
`["/home/user/.pyenv/versions/miniconda3-latest/envs/ahc/bin/python" "-m" "ipykernel_launcher" "-f" "{connection_file}"]`,

which both looks good I guess.
I also tried adding to my init file, as @sjkuntz  recommended in the last post, but that didn't change anything. Is there anything I could have configured wrongly?

**Comment by magonreal** [comment]
I guess the reason for this is that my `module-file-suffix` is `nil`. So my emacs installation has no dynamic modules support. I will look into this.

**Comment by magonreal** [comment]
Now I installed emacs from source with `./configure --with-modules` and it's working perfectly. The pyenv virtualenv is also making no problems on my side.

**Comment by nanounanue** [comment]
Hi!

Sorry for taking so long to answer, I am _still_ getting an error:

```
executing Jupyter-Python code block...
jupyter--block-until-conn-file-access: Kernel process exited:
/usr/bin/python: No module named ipykernel_launcher

Process jupyter-kernel-python3 exited abnormally with code 1
```

I am using virtual environments. Printing the version of `python` in the directory I get:

```
$ which python                   
/home/nanounanue/.pyenv/shims/python
```

This is different that the one `emacs-jupyter` tries to use

**Comment by magonreal** [comment]
And your `pyenv` is activating normally when you open that directory in the terminal?
Recently, on one computer, I had some problems with `pyenv` not activating by itself when I go to a directory with a `.python-version`

**Comment by nanounanue** [comment]
Thank you for your quick response!

Yes, `pyenv` is activating the environment when I enter to that directory
using the command line.

But, emacs wasn't started in that directory thought

A.

**Comment by magonreal** [comment]
How about python functionality outside of `emacs-jupyter` ?
For example, the flake8 checking for python documents as described in https://realpython.com/emacs-the-best-python-editor/

If this also does not work, then I guess this is not a problem with `emacs-jupyter`, but some other emacs related problem instead.

**Comment by dzop** [comment]
@nanounanue  are you using https://github.com/jorgenschaefer/pyvenv?

**Comment by magonreal** [comment]
Ah. I am using spacemacs here..
So maybe the setup of `pyenv` is a bit different for me (you don't have to configure anything, it just works out of the box for spacemacs)

**Comment by Yevgnen** [comment]
I would suggest use `virtualenvwrapper.el`, sample config:

```elisp
(use-package virtualenvwrapper
  :ensure t
  :init (setq venv-location (expand-file-name "~/.pyenv/versions"))
  :config (venv-initialize-interactive-shells))
```

and the `.dir-locals.el` inside my directory of `org` files

```elisp
;;; Directory Local Variables
;;; For more information see (info "(emacs) Directory Variables")

((org-mode . ((eval . (venv-workon "some-env")))))
```

**Comment by nanounanue** [comment]
Thank you for your suggestions, I will try it as soon as I find a fix for #160

**Comment by dzop** [comment]
@nanounanue any luck on fixing this now that you solved #160?

**Comment by dzop** [comment]
Closing since the method mentioned in https://github.com/dzop/emacs-jupyter/issues/93#issuecomment-512899086 appears to work for others.

Let me know if there are still issues.
